### PR TITLE
feat(GraphQL): Webhooks on add/update/delete mutations (GRAPHQL-1045)

### PR DESCRIPTION
**Pull request by abhimanyusinghgaur**
Please see the [RFC](https://discuss.dgraph.io/t/webhook-lambda-on-add-update-delete-mutations/12904) for more details.

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7494)

### CODE REVIEW

**Review by pawanrawal — APPROVED** [review]
> Reviewed 74 of 74 files at r1.
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7494)* status: all files reviewed, 3 unresolved discussions (waiting on @abhimanyusinghgaur, @manishrjain, @vmrajas, and @vvbalaji-dgraph)

___
*[graphql/resolve/mutation_rewriter.go, line 819 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7494#-MUlf9ma4R02g2sYxc1x:-MUlf9mb5VMR5xEkQffz:b7q7rm1) ([raw file](https://github.com/dgraph-io/dgraph/blob/446a15fcd39a3be86be289356d3aae91d8c0a1ad/graphql/resolve/mutation_rewriter.go#L819)):*
> ```Go
> 	uids := make([]uint64, 0, len(uidSlice))
> 
> 	if len(uidSlice) > 0 {
> ```

this if block is not required
___
*[graphql/resolve/webhook.go, line 110 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7494#-MUlflUS6V9mvOaxszPG:-MUlflUS6V9mvOaxszPH:b-3fn0r2) ([raw file](https://github.com/dgraph-io/dgraph/blob/446a15fcd39a3be86be289356d3aae91d8c0a1ad/graphql/resolve/webhook.go#L110)):*
> ```Go
> 	b, err := json.Marshal(payload)
> 	if err != nil {
> 		// don't care to send the payload if there are JSON marshalling errors
> ```

We should log the error using glog.Error in V3 level
___
*[graphql/resolve/webhook.go, line 117 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7494#-MUlg2Ea4S4Z7cDgEJ_g:-MUlg2Eb16keqPUHDVEb:b6keuro) ([raw file](https://github.com/dgraph-io/dgraph/blob/446a15fcd39a3be86be289356d3aae91d8c0a1ad/graphql/resolve/webhook.go#L117)):*
> ```Go
> 	headers.Set("Content-Type", "application/json")
> 	// don't care for the response/errors, if any.
> 	_, _ = schema.MakeHttpRequest(nil, http.MethodPost, x.Config.GraphQL.GetString("lambda-url"),
> ```

let's log the error if its non-nil or if response code is not a success

**Review by pawanrawal — COMMENTED** [review]
Also add a way to override this behavior using a key that is part of extensions.

> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7494#-:-MUlhFuM0KNMeoa-l7DK:b-wgah7d)* status: all files reviewed, 3 unresolved discussions (waiting on @abhimanyusinghgaur, @manishrjain, @vmrajas, and @vvbalaji-dgraph)

File: graphql/e2e/common/common.go [hunk 1]
```diff
@@ -797,8 +800,8 @@ func RunAll(t *testing.T) {
 	t.Run("query only typename", queryOnlyTypename)
 	t.Run("query nested only typename", querynestedOnlyTypename)
 	t.Run("test onlytypename for interface types", onlytypenameForInterface)
-	t.Run("entitites Query on extended type with key field of type String", entitiesQueryWithKeyFieldOfTypeString)
-	t.Run("entitites Query on extended type with key field of type Int", entitiesQueryWithKeyFieldOfTypeInt)
+	t.Run("entities Query on extended type with key field of type String", entitiesQueryWithKeyFieldOfTypeString)
+	t.Run("entities Query on extended type with key field of type Int", entitiesQueryWithKeyFieldOfTypeInt)
```
Comment (vmrajas) [inline graphql/e2e/common/common.go:804]: :+1:

File: graphql/e2e/directives/schema.graphql [hunk 2]
```diff
@@ -341,7 +341,8 @@ type Region {
     district: District
 }
 
-type District {
+type District @lambdaOnMutate(add: true, update: true, delete: true) {
+	dgId: ID!
```
Comment (vmrajas) [inline graphql/e2e/directives/schema.graphql:345]: nit: Indentation seems off. Likely due to different tab spaces being used.

File: graphql/e2e/directives/script.js [hunk 3]
```diff
@@ -52,4 +52,19 @@ async function rank({parents}) {
...
+        body: JSON.stringify(event)
+    })
+    // just return, nothing else to do with response
+}
+
+self.addWebHookResolvers({
+    "District.add": districtWebhook,
+    "District.update": districtWebhook,
+    "District.delete": districtWebhook,
 })
```
Comment (vmrajas) [inline graphql/e2e/directives/script.js:70]: nit: Line at end of file.

File: graphql/e2e/normal/script.js [hunk 4]
```diff
@@ -52,4 +52,23 @@ async function rank({parents}) {
...
+        body: JSON.stringify(event)
+    })
+    // just return, nothing else to do with response
+}
+
+self.addWebHookResolvers({
+    "District.add": districtWebhook,
+    "District.update": districtWebhook,
+    "District.delete": districtWebhook,
 })
```
Comment (vmrajas) [inline graphql/e2e/normal/script.js:74]: nit: Extra line at end of file.

File: graphql/resolve/mutation.go [hunk 5]
```diff
@@ -215,9 +220,9 @@ func (mr *dgraphResolver) rewriteAndExecute(
 	defer func() {
 		if !commit && mutResp != nil && mutResp.Txn != nil {
 			mutResp.Txn.Aborted = true
-			err := mr.executor.CommitOrAbort(ctx, mutResp.Txn)
+			_, err := mr.executor.CommitOrAbort(ctx, mutResp.Txn)
 			if err != nil {
-				glog.Errorf("Error occured while aborting transaction: %s", err)
+				glog.Errorf("Error occurred while aborting transaction: %s", err)
```
Comment (vmrajas) [inline graphql/resolve/mutation.go:225]: :+1:

File: graphql/resolve/mutation_rewriter.go [hunk 6]
```diff
@@ -263,12 +263,12 @@ func (xidMetadata *xidMetadata) isDuplicateXid(atTopLevel bool, xidVar string,
 // simply ignored.
 // If it is found out that the Person with id 0x123 does not exist, the corresponding
 // mutation will fail.
-func (mrw *AddRewriter) RewriteQueries(
+func (arw *AddRewriter) RewriteQueries(
```
Comment (vmrajas) [inline graphql/resolve/mutation_rewriter.go:266]: :+1:

File: graphql/schema/gqlschema_test.yml [hunk 7]
```diff
@@ -3316,3 +3344,16 @@ valid_schemas:
...
+    input: |
+      interface Post @lambdaOnMutate(add: true, delete: false) {
+        id: ID!
+        title: String
+      }
+
+      type Question implements Post @lambdaOnMutate(add: true, update: true) {
+        id: ID!
+        questionText: String
+      }
```
Comment (vmrajas) [inline graphql/schema/gqlschema_test.yml:3359]: nit: Line at end of file

**Review by vmrajas — APPROVED** [review]
LGTM
The mutation_rewriter.go part looks good to me.
Just added few nits.

**Review by abhimanyusinghgaur — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/7494)* status: 54 of 91 files reviewed, 6 unresolved discussions (waiting on @manishrjain, @pawanrawal, @vmrajas, and @vvbalaji-dgraph)

___
*[graphql/resolve/mutation_rewriter.go, line 819 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7494#-MUlf9ma4R02g2sYxc1x:-MW7xWRA55ah9BETV--j:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/446a15fcd39a3be86be289356d3aae91d8c0a1ad/graphql/resolve/mutation_rewriter.go#L819)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

this if block is not required
</blockquote></details>

Done.
___
*[graphql/resolve/webhook.go, line 110 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7494#-MUlflUS6V9mvOaxszPG:-MW7xYZk6AvOMPhxzqQ9:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/446a15fcd39a3be86be289356d3aae91d8c0a1ad/graphql/resolve/webhook.go#L110)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

We should log the error using glog.Error in V3 level
</blockquote></details>

Done.
___
*[graphql/resolve/webhook.go, line 117 at r1](https://reviewable.io/reviews/dgraph-io/dgraph/7494#-MUlg2Ea4S4Z7cDgEJ_g:-MW7xaV4-sMDKtyL9rk0:b-896fix) ([raw file](https://github.com/dgraph-io/dgraph/blob/446a15fcd39a3be86be289356d3aae91d8c0a1ad/graphql/resolve/webhook.go#L117)):*
<details><summary><i>Previously, pawanrawal (Pawan Rawal) wrote…</i></summary><blockquote>

let's log the error if its non-nil or if response code is not a success
</blockquote></details>

Done.

**Pull request merged**